src/main/java/graphql/util/TraverserState.java

```diff
@@ -89,6 +89,10 @@ public void clear() {
         visited.clear();
     }
 
+    public void addVisited(T visited) {
+        this.visited.add(visited);
```

Comment:
any particular reason why visited set is declared as `LInkedHashSet`?
Declaration like this could create false-positive results for nodes that implement equals and hashCode methods.
Cycle detection should prevent visiting exactly same node again, not the node that equals to it, cause that could be an absolutely legitimate subtree.

Reply:
I would consider that a problem of the user of the Traverser: If you implement equals/hashCode and have two different Nodes in your Tree which equals you made something wrong.
And the only user of the Traverser is graphql-java itself, this means we are responsible for that.